pkg/providers/instance/vminstance.go

```diff
@@ -642,11 +643,25 @@ func (p *DefaultVMProvider) createVirtualMachine(ctx context.Context, opts *crea
 	}
 	vm := newVMObject(opts)
 	log.FromContext(ctx).V(1).Info("creating virtual machine", "vmName", opts.VMName, logging.InstanceType, opts.InstanceType.Name)
-	metrics.VMCreateStartMetric.Inc(ctx, "creating virtual machine", metrics.ImageID(opts.LaunchTemplate.ImageID))
+	VMCreateStartMetric.With(map[string]string{
+		metrics.ImageLabel:        opts.LaunchTemplate.ImageID,
+		metrics.SizeLabel:         opts.InstanceType.Name,
+		metrics.ZoneLabel:         opts.Zone,
+		metrics.CapacityTypeLabel: opts.CapacityType,
+		metrics.NodePoolLabel:     opts.NodePoolName,
+	}).Inc()
 
 	poller, err := p.azClient.virtualMachinesClient.BeginCreateOrUpdate(ctx, p.resourceGroup, opts.VMName, *vm, nil)
 	if err != nil {
-		metrics.VMCreateSyncFailureMetric.Inc(ctx, "failed to create virtual machine on initial put", metrics.ImageID(opts.LaunchTemplate.ImageID))
+		VMCreateFailureMetric.With(map[string]string{
+			metrics.ImageLabel:        opts.LaunchTemplate.ImageID,
+			metrics.SizeLabel:         opts.InstanceType.Name,
+			metrics.ZoneLabel:         opts.Zone,
+			metrics.CapacityTypeLabel: opts.CapacityType,
+			metrics.NodePoolLabel:     opts.NodePoolName,
+			metrics.PhaseLabel:        phaseSyncFailure,
+			metrics.ErrorCodeLabel:    err.Error(),
```

Comment:
I don't think this is actually the error_code we want, it's just going to be a  big string with a lot of possible values and blow out cardinality, isn't it?

I think we need to extract the error_code from the ARM HTTP response. See https://github.com/Azure/azure-sdk-for-go-extensions/blob/69ee2f9f2b8a9610e0d1f31738f364079b6cc262/pkg/errors/armerrors.go#L50C11-L50C26

I think we need to find the AzureError code and put that here. If we can't find the code we can fall back to Unknown.

Reply:
Same applies on the async side (the error parsing should be the same)